File: Codeforces/1918D.cpp
#include <bits/stdc++.h>
using namespace std;

#define int long long
const int INF = 1e18;

class Window {
public:
    int l, r;
    deque<pair<int, int>> deq;

    Window() {
        l = 0;
        r = 0;
    }

    void pop() {
        if (deq.front().second == l)
            deq.pop_front();
        l++;
    }

    void append(int x) {
        while (!deq.empty() && deq.back().first >= x)
            deq.pop_back();
        deq.push_back({x, r});
        r++;
    }

    int get_min() {
        return deq.empty() ? INF : deq.front().first;
    }
};

void solve() {
    int n;
    cin >> n;
    vector<int> a(n + 1);
    for (int i = 1; i <= n; ++i)
        cin >> a[i];
    vector<int> pre(n + 1);
    partial_sum(a.begin(), a.end(), pre.begin());

    auto check = [&](int k) {
        vector<int> dp(n + 1, INF);
        dp[0] = 0;

        int j = 0;
        Window w;
        w.append(0);
        for (int i = 1; i <= n; ++i) {
            while (pre[i - 1] - pre[j] > k) {
                w.pop();
                j++;
            }

            dp[i] = a[i] + w.get_min();
            w.append(dp[i]);
        }

        int res = INF;
        int x = 0;
        for (int i = n; i >= 1; --i) {
            res = min(res, dp[i]);
            x += a[i];
            if (x > k)
                break;
        }

        return res <= k;
    };

    int lo = 1, hi = 1e14;
    while (lo < hi) {
        int m = lo + (hi - lo) / 2;
        if (check(m))
            hi = m;
        else
            lo = m + 1;
    }
    cout << lo << endl;
}

signed main() {
    int t;
    cin >> t;
    while (t--)
        solve();
    return 0;
}

File: Codeforces/1918D.py
from collections import deque
from itertools import accumulate

class Window:
    def __init__(self):
        self.l = 0
        self.r = 0
        self.deq = deque()

    def pop(self):
        if self.deq[0][1] == self.l:
            self.deq.popleft()
        self.l += 1

    def append(self, x):
        while self.deq and self.deq[-1][0] >= x:
            self.deq.pop()
        self.deq.append((x, self.r))
        self.r += 1

    def get_min(self):
        return self.deq[0][0] if self.deq else float('inf')

def solve():
    n = int(input())
    a = list(map(int, input().split()))
    a.insert(0, 0)
    pre = list(accumulate(a))

    # whether ans <= k
    def check(k):
        # number taken -> min sum
        dp = [float('inf')]*(n+1)
        dp[0] = 0
        
        j = 0
        w = Window()
        w.append(0)
        for i in range(1, n+1):
            while pre[i-1]-pre[j] > k:
                w.pop()
                j += 1

            dp[i] = a[i]+w.get_min()
            w.append(dp[i])


        res = float('inf')
        x = 0
        for i in range(n, 0, -1):
            res = min(res, dp[i])
            x += a[i]
            if x > k:
                break


        return res <= k
            

    lo = 1
    hi = int(1e14)
    while lo < hi:
        m = (lo+hi)//2
        if check(m):
            hi = m
        else:
            lo = m+1
    print(lo)


t = int(input())
for _ in range(t):
    solve()
